samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java

```diff
@@ -144,122 +58,108 @@ public StreamGraphImpl(ApplicationRunner runner, Config config) {
+   * @param keyExtractor the function to extract the output key from the input message
+   * @param msgExtractor the function to extract the output message from the input message
+   * @param msgBuilder the function to convert the incoming key and message to the outgoing message of type {@code M}
+   * @param keySerde  the serde used to serialize/deserialize the message key from the intermediate {@link MessageStream}
+   * @param msgSerde  the serde used to serialize/deserialize the message body from the intermediate {@link MessageStream}
+   * @param <K> the type of key in the incoming message
+   * @param <V> the type of message in the incoming message
+   * @param <M> the type of messages in the intermediate {@link MessageStream}
+   * @return the intermediate {@link MessageStream} object
    */
-  @Override
-  public <K, V, M extends MessageEnvelope<K, V>> OutputStream<M> createIntStream(StreamSpec streamSpec,
+  <K, V, M> MessageStreamImpl<M> createIntStream(StreamSpec streamSpec,
+      Function<M, K> keyExtractor, Function<M, V> msgExtractor, BiFunction<K, V, M> msgBuilder,
       Serde<K> keySerde, Serde<V> msgSerde) {
-    if (!this.inStreams.containsKey(streamSpec.getId())) {
-      this.inStreams.putIfAbsent(streamSpec.getId(), new IntermediateStreamImpl<K, K, V, M>(this, streamSpec, keySerde, msgSerde));
-    }
-    IntermediateStreamImpl<K, K, V, M> intStream = (IntermediateStreamImpl<K, K, V, M>) this.inStreams.get(streamSpec.getId());
-    if (!this.outStreams.containsKey(streamSpec.getId())) {
-      this.outStreams.putIfAbsent(streamSpec.getId(), intStream);
-    }
+    IntermediateStreamImpl<K, V, M> intStream = (IntermediateStreamImpl<K, V, M>) inStreams.computeIfAbsent(
+        streamSpec.getId(), streamId -> new IntermediateStreamImpl<>(this, streamSpec,
+            keyExtractor, msgExtractor, msgBuilder, keySerde, msgSerde));
+    outStreams.putIfAbsent(streamSpec.getId(), intStream);
     return intStream;
   }
 
-  @Override public Map<StreamSpec, MessageStream> getInStreams() {
+  public Map<StreamSpec, MessageStream> getInStreams() {
     Map<StreamSpec, MessageStream> inStreamMap = new HashMap<>();
-    this.inStreams.forEach((ss, entry) -> inStreamMap.put(((InputStreamImpl) entry).getSpec(), entry));
-    return Collections.unmodifiableMap(inStreamMap);
+    this.inStreams.forEach((ss, entry) -> inStreamMap.put(entry.getStreamSpec(), entry));
+    return inStreamMap;
   }
 
-  @Override public Map<StreamSpec, OutputStream> getOutStreams() {
-    Map<StreamSpec, OutputStream> outStreamMap = new HashMap<>();
-    this.outStreams.forEach((ss, entry) -> {
-        StreamSpec streamSpec = (entry instanceof IntermediateStreamImpl) ?
-          ((IntermediateStreamImpl) entry).getSpec() :
-          ((OutputStreamImpl) entry).getSpec();
-        outStreamMap.put(streamSpec, entry);
-      });
-    return Collections.unmodifiableMap(outStreamMap);
+  public Map<StreamSpec, MessageStream> getOutStreams() {
```

Comment:
Yeah I was thinking about this too. This OutputOperatorSpec would be the same as the SinkOperatorSpec. The InputOperatorSpec is the (currently missing) spec for the RootOperatorImpl. Both are better names for the operators. Let me try this out and see if the ExecutionPlanner changes required for this make sense too.